pkg/tls/reloader_test.go

```diff
@@ -0,0 +1,376 @@
+		require.NoError(t, os.WriteFile(certFile, certPEM1, 0o600))
+		require.NoError(t, os.WriteFile(keyFile, keyPEM1, 0o600))
+
+		log := logger.GetLogger("tls-test")
+		reloader, err := NewReloader(certFile, keyFile, log)
+		require.NoError(t, err)
+		defer reloader.Stop()
+
+		// Start reloader
+		err = reloader.Start()
+		require.NoError(t, err)
+
+		// Wait for initial certificate to be loaded
+		time.Sleep(100 * time.Millisecond)
+
+		// Verify initial certificate
+		tlsConfig := reloader.GetTLSConfig()
+		cert, err := tlsConfig.GetCertificate(nil)
+		require.NoError(t, err)
+		leafCert, err := x509.ParseCertificate(cert.Certificate[0])
+		require.NoError(t, err)
+		require.Equal(t, "test1.local", leafCert.Subject.CommonName)
+
+		// Generate and replace with new certificate files
+		certPEM2, keyPEM2 := generateSelfSignedCert(t, "test2.local")
+		require.NoError(t, os.WriteFile(certFile, certPEM2, 0o600))
+		require.NoError(t, os.WriteFile(keyFile, keyPEM2, 0o600))
+
+		// Wait for watcher to detect changes
+		time.Sleep(500 * time.Millisecond)
+
+		// Verify certificate has been updated
+		cert, err = tlsConfig.GetCertificate(nil)
+		require.NoError(t, err)
+		leafCert, err = x509.ParseCertificate(cert.Certificate[0])
+		require.NoError(t, err)
+		require.Equal(t, "test2.local", leafCert.Subject.CommonName)
+	})
+
+	// Test removing old files and creating new files
+	t.Run("removing and creating new files", func(t *testing.T) {
+		tempDir := t.TempDir()
+		certFile := filepath.Join(tempDir, "cert.pem")
+		keyFile := filepath.Join(tempDir, "key.pem")
+
+		// Create initial certificate
+		certPEM1, keyPEM1 := generateSelfSignedCert(t, "test1.local")
+		require.NoError(t, os.WriteFile(certFile, certPEM1, 0o600))
+		require.NoError(t, os.WriteFile(keyFile, keyPEM1, 0o600))
+
+		log := logger.GetLogger("tls-test")
+		reloader, err := NewReloader(certFile, keyFile, log)
+		require.NoError(t, err)
+		defer reloader.Stop()
+
+		// Start reloader
+		err = reloader.Start()
+		require.NoError(t, err)
+
+		// Wait for initial certificate to be loaded
+		time.Sleep(100 * time.Millisecond)
+
+		// Remove the files
+		require.NoError(t, os.Remove(certFile))
+		require.NoError(t, os.Remove(keyFile))
+		time.Sleep(100 * time.Millisecond)
+
+		// Create new files with different content
+		certPEM2, keyPEM2 := generateSelfSignedCert(t, "test3.local")
+		require.NoError(t, os.WriteFile(certFile, certPEM2, 0o600))
+		require.NoError(t, os.WriteFile(keyFile, keyPEM2, 0o600))
+		time.Sleep(500 * time.Millisecond)
+
+		// Verify the certificates have been updated
+		tlsConfig := reloader.GetTLSConfig()
+		cert, err := tlsConfig.GetCertificate(nil)
+		require.NoError(t, err)
+		leafCert, err := x509.ParseCertificate(cert.Certificate[0])
+		require.NoError(t, err)
+		require.Equal(t, "test3.local", leafCert.Subject.CommonName)
+	})
+
+	// Test creating files after reloader starts
+	t.Run("creating files after reloader starts", func(t *testing.T) {
```

Comment:
The reloader can't wait for the absent files. I think that's fine for the use case. Let's delete this test case.